# Lecture 5 — Tutorials 3.1–3.14 Exercises

**Goal.** Attempt All 41 numbered website exercises before opening the clearly separated solution section.

This notebook matches the numbered Lecture 5 website tutorials. Examples and exercises share one continuous sequence within each tutorial, so these exercise numbers continue after the examples. The dataset is synthetic, contains deliberate quality problems and must not be interpreted as evidence about real cities or services.

**Local setup.** Google Colab includes pandas. On a local machine, run `python -m pip install -r requirements.txt` from the repository, or install pandas with `python -m pip install "pandas>=2.0"`.


## Setup

Run this cell first. It loads the versioned raw CSV from GitHub and checks the expected 121 × 14 structure. Keep the raw data unchanged during Lecture 5.


In [1]:
import pandas as pd  # Loads the library feature used by this step.
data_url = "https://raw.githubusercontent.com/asmrabbi/E26_TAN7_Scripting_CPH/main/data/E26_TAN7_service_experience_raw.csv"  # Stores the calculated or supplied value for later use.
df = pd.read_csv(data_url)  # Stores the calculated or supplied value for later use.
print("Dataset shape:", df.shape)  # Displays the result so it can be inspected.
assert df.shape == (121, 14), "The wrong dataset version was loaded."  # Stores the calculated or supplied value for later use.
display(df.head())  # Displays the result so it can be inspected.


Dataset shape: (121, 14)
     record_id  ...                                           feedback
0  SR2026-0001  ...  The guidance was clear, but the supporting-doc...
1  SR2026-0002  ...  Several users asked for clearer information ab...
2  SR2026-0003  ...  Useful advice and a friendly response from staff.
3  SR2026-0004  ...  Helpful guidance, but some users reported dela...
4  SR2026-0005  ...  The response was fast and the information was ...

[5 rows x 14 columns]


# Student exercises

Work in website order. Write explanations in Markdown cells and code in the starter cells. Do not scroll to the solution divider until you have made an attempt.


## Exercise 3.1.5 — Describe a dataset before coding

A fictional dataset contains these columns:

```text
organisation_id, year, sector, employees, energy_use_kwh, sustainability_comment
```

Answer without writing Python:

1. What might one row represent?
2. Which columns look numeric?
3. Which column looks like an identifier?
4. Which column contains free text?
5. What information would you need before interpreting `energy_use_kwh`?
6. Why would the number of rows not necessarily equal the number of organisations?


In [2]:
pass  # Runs this part of the documented analysis.


## Exercise 3.2.4 — Find the structural problem

Consider:

```csv
id;city;score
1;Copenhagen;4.2
2;Aalborg;3.9
```

A student runs:

```python
df = pd.read_csv("scores.csv")
```

and gets one column instead of three.

1. What is the likely problem?
2. What argument should the student investigate?
3. Why should they inspect the file before changing the code?


In [3]:
pass  # Runs this part of the documented analysis.


## Exercise 3.2.5 — Missing does not mean zero

Explain why these two values should not automatically be treated as equivalent:

```text
cases_received = 0
cases_received = blank
```


In [4]:
pass  # Runs this part of the documented analysis.


## Exercise 3.3.14 — Load and verify

1. Import pandas as `pd`.
2. Load the course CSV from the raw GitHub URL.
3. Display the first five rows.
4. Store the DataFrame in a variable called `df`.
5. Use `type(df)` and explain the result.


In [5]:
pass  # Runs this part of the documented analysis.


## Exercise 3.3.15 — Explain the file/DataFrame distinction

A student says:

> “I changed a value in `df`, so I have permanently changed the original CSV on GitHub.”

Explain why this is normally false.


In [6]:
pass  # Runs this part of the documented analysis.


## Exercise 3.4.16 — Produce an inspection note

Using the course DataFrame, write code that reports:

1. number of rows;
2. number of columns;
3. all column names;
4. first three rows;
5. last three rows;
6. technical dtypes;
7. `df.info()`.

Then write three sentences describing what you learned **without making claims about real service performance**.


In [7]:
pass  # Runs this part of the documented analysis.


## Exercise 3.4.17 — Why inspect beyond the first five rows?

Explain two reasons why `df.head()` alone is not enough for data-quality assessment.


In [8]:
pass  # Runs this part of the documented analysis.


## Exercise 3.5.21 — Column selection

Write code to display only:

- `record_id`;
- `service_type`;
- `satisfaction_score`;
- `feedback`.

Then explain whether the result is a Series or DataFrame.


In [9]:
pass  # Runs this part of the documented analysis.


## Exercise 3.5.22 — Filter a numeric condition

Find rows where `cases_received` is at least 100.


In [10]:
pass  # Runs this part of the documented analysis.


## Exercise 3.5.23 — Combine a category and numeric condition

Find Transport records with more than 90 cases received. Display only:

- `record_id`;
- `city`;
- `cases_received`.


In [11]:
pass  # Runs this part of the documented analysis.


## Exercise 3.5.24 — Housing reports for review

You are helping a municipal research team prepare a first review of Housing service reports. The team wants to inspect months with more than 120 received cases. Use the raw teaching CSV and keep the original table available.

1. Select reports whose service_type is exactly "Housing" and whose cases_received is greater than 120.
2. Show city and cases_received for the first three matching reports.
3. Print the total number of matching reports before limiting the display.
4. Write one sentence explaining what that number counts and one limitation of the selection.

Predict the output, run your code, then explain which rows the rule keeps.


In [ ]:
# Write your solution here.  # Write and test your own solution before opening the solution section.


## Exercise 3.6.4 — Classify the course variables

Create a table with these columns:

```text
variable | conceptual_type | expected_values_or_unit | possible_quality_question
```

Complete it for all fourteen variables.

Do not copy `df.dtypes` as the conceptual type.


In [12]:
pass  # Runs this part of the documented analysis.


## Exercise 3.6.5 — Explain why this statement is wrong

> “`record_id` is an integer, so calculating its average is useful.”


In [13]:
pass  # Runs this part of the documented analysis.


## Exercise 3.7.15 — Produce a numeric summary

For `cases_resolved`, calculate:

1. count;
2. sum;
3. mean;
4. median;
5. minimum;
6. maximum;
7. standard deviation;
8. `describe()`.

Then write two observations and one quality question.


In [14]:
pass  # Runs this part of the documented analysis.


## Exercise 3.7.16 — Mean or median?

Calculate the mean and median of `cases_received`. Explain:

1. why they differ;
2. why the negative raw value matters;
3. why neither statistic alone tells you whether the dataset is reliable.


In [15]:
pass  # Runs this part of the documented analysis.


## Exercise 3.7.17 — Count the denominator explicitly

Calculate the mean satisfaction score and also print how many non-missing satisfaction scores contributed to the calculation.


In [16]:
pass  # Runs this part of the documented analysis.


## Exercise 3.8.10 — Derive a difference

Create:

```text
case_gap = cases_received - cases_resolved
```

Then:

1. show the five smallest values;
2. identify any negative results;
3. explain why a negative result requires investigation.


In [17]:
pass  # Runs this part of the documented analysis.


## Exercise 3.8.11 — Explain two rate formulas

Write code for:

1. the mean of row-level resolution percentages;
2. total resolved divided by total received.

Then explain in ordinary language why they differ.


In [18]:
pass  # Runs this part of the documented analysis.


## Exercise 3.8.12 — Calculate a missing percentage

Calculate the percentage of missing values in `satisfaction_score` using:

```text
missing count / total rows × 100
```

Expected raw-data result:

```text
3.31
```


In [19]:
pass  # Runs this part of the documented analysis.


## Exercise 3.9.14 — Frequencies and percentages

For `service_type`:

1. print unique values;
2. count unique values;
3. create a frequency table;
4. create a percentage table;
5. identify the most frequent stored category.


In [20]:
pass  # Runs this part of the documented analysis.


## Exercise 3.9.15 — Grouped summary

Calculate `count`, `mean`, `median`, `min` and `max` of `cases_received` for each `service_type`.

Then answer:

1. Which group contains the suspicious negative minimum?
2. Why might that affect the mean?
3. Why should the output still be labelled preliminary?


In [21]:
pass  # Runs this part of the documented analysis.


## Exercise 3.9.16 — Discover category fragmentation

Group `cases_received` by raw `city` and inspect the output.

Write a short note explaining why the current city groups should probably be reviewed before final analysis.


In [22]:
pass  # Runs this part of the documented analysis.


## Exercise 3.9.17 — Create a crosstab

Create a cross-tabulation of `city` by `service_type`.

Then explain:

- what one cell counts;
- why the raw city inconsistencies matter;
- what a row-normalised percentage would mean.


In [23]:
pass  # Runs this part of the documented analysis.


## Exercise 3.10.19 — Build a quality audit

Write code that reports:

1. missing values by column;
2. missing percentages;
3. exact duplicate count;
4. all rows involved in exact duplication;
5. repeated `record_id` values;
6. unique raw city labels;
7. negative `cases_received` rows;
8. out-of-range satisfaction rows;
9. rows where resolved exceeds received;
10. non-numeric `resolution_days` values;
11. invalid `report_month` values.

For each detected issue, write **one sentence describing the issue without fixing it**.


In [24]:
pass  # Runs this part of the documented analysis.


## Exercise 3.10.20 — Classify the quality issue

For each item below, classify it mainly as completeness, uniqueness, validity, consistency, plausibility or logical consistency. More than one answer may sometimes be defensible.

1. `feedback` is blank.
2. `record_id SR2026-0027` is repeated in identical rows, while `SR2026-0079` is reused for two different rows.
3. `city = " copenhagen "` appears beside `city = "Copenhagen"`.
4. `satisfaction_score = 0.0` or `6.2` when the documented scale is 1–5.
5. `cases_resolved = 90` and `cases_received = 85` under an assumption that resolved cases must be a subset of received cases.
6. `report_month = 2026-13-01`.


In [25]:
pass  # Runs this part of the documented analysis.


## Exercise 3.10.21 — checking a satisfaction claim

A research team is drafting a paragraph about satisfaction in the same dataset. A colleague has quoted the average, but the teaching dictionary allows scores from 1 to 5. Before using that number, the team needs to check for missing and out-of-range scores.

1. Count missing satisfaction scores.
2. Find scores below 1 or above 5. Display record_id and the stored score for those reports.
3. Print the raw mean and the number of recorded scores contributing to it.
4. Write a cautious interpretation stating the result, the quality concerns and what should happen next.

Keep the stored values. Detecting a concern does not determine its correction.


In [ ]:
# Write your solution here.  # Write and test your own solution before opening the solution section.


## Exercise 3.11.14 — Explore text before coding themes

1. Display all non-missing feedback comments.
2. Read them manually.
3. Write down at least three recurring ideas you notice.
4. For each idea, identify possible words that could help retrieve relevant comments.
5. Explain at least one way that your keyword rule could miss relevant wording.


In [26]:
pass  # Runs this part of the documented analysis.


## Exercise 3.11.15 — Search and validate a keyword rule

Create a case-insensitive search for comments containing either `quick` or `fast`.

Then:

1. count matches;
2. display the matching full comments;
3. read each match;
4. decide whether each one genuinely concerns speed;
5. identify one possible synonym your search rule would miss.


In [27]:
pass  # Runs this part of the documented analysis.


## Exercise 3.11.16 — Compare quantitative and qualitative evidence

Inspect the comments that match `wait|long` together with `satisfaction_score`.

Write:

- one observation directly supported by the rows;
- one cautious interpretation;
- one claim that would be too strong.


In [28]:
pass  # Runs this part of the documented analysis.


## Exercise 3.12.9 — Design a coding rule

Choose one idea from the feedback, for example:

- clarity;
- usefulness/helpfulness;
- speed;
- difficulty.

Then:

1. define the idea in one sentence;
2. choose one or more search terms;
3. create a Boolean indicator;
4. count `True` values;
5. display all matching comments;
6. manually inspect at least three non-matching comments;
7. identify one possible false positive and one possible false negative;
8. revise the rule if justified;
9. document the final rule.


In [29]:
pass  # Runs this part of the documented analysis.


## Exercise 3.12.10 — Compare code and score

Using your Boolean indicator:

```python
df.groupby("your_indicator")["satisfaction_score"].agg(
    ["count", "mean", "median"]
)
```

Write:

- one direct observation;
- one cautious interpretation;
- one causal claim that the result does **not** justify.


In [30]:
pass  # Runs this part of the documented analysis.


## Exercise 3.12.11 — Multiple codes per comment

Create at least two indicators and find comments where both are `True`.

Example structure:

```python
both = df[
    df["mentions_helpful"] &
    df["mentions_wait"]
]

display(both[["feedback"]])
```

Explain why allowing several codes can be more faithful to mixed comments than assigning only one category.


In [31]:
pass  # Runs this part of the documented analysis.


## Exercise 3.13.3 — Create a data dictionary

Create a Markdown table for all fourteen variables with:

```text
variable
meaning
conceptual type
unit/scale
expected values
known/raw concern
```


In [32]:
pass  # Runs this part of the documented analysis.


## Exercise 3.13.4 — Rewrite an overclaim

Rewrite this statement so it becomes defensible:

> “Transport has the highest cases, so public transport is the biggest problem in Danish cities.”


In [33]:
pass  # Runs this part of the documented analysis.


## Exercise 3.13.5 — Separate three levels of statement

For one calculation from Tutorial 3.9, write:

1. one direct observation;
2. one cautious interpretation;
3. one claim that the dataset cannot support.


In [34]:
pass  # Runs this part of the documented analysis.


# Tutorial 3.14 integrated exercises

Each exercise below starts a new situation. Write and test your own solution before using the solution section.


## Exercise 3.14.9 — Prepare a two-city service briefing

A TAN research team is preparing a short briefing about service records from Copenhagen and Aalborg. The team has received the raw teaching dataset and knows that one row is intended to describe one service type in one city during one reporting month. A coordinator wants a quick comparison of workload, resolved cases and satisfaction across the two cities. Before calculating anything, the team needs proof that the correct file was loaded and that the relevant rows were selected. The raw city column also contains spelling, capitalisation and spacing variants that could split records into misleading groups. Missing satisfaction values mean that the number of rows and the number of scores will not always be the same. Written feedback may add useful context, but a keyword match cannot represent every resident experience. Your briefing must therefore combine inspection, filtering, grouped calculations, a text check and a clear limitation statement.

### Your task

1. Load the raw CSV, confirm its shape and display the first five rows.
2. Inspect every raw city label and its frequency before selecting data.
3. Select exact `Copenhagen` and `Aalborg` records and display only the identifier, month, city, service type, cases, satisfaction and feedback columns.
4. For each selected city, calculate row count, total and mean cases received, total cases resolved, and the count, mean and median of satisfaction scores.
5. Create a transparent waiting-related keyword rule, count its matches by city and display the matching comments.
6. Write three observations and explain how city-label variants, missing scores and the synthetic source limit the briefing.


In [ ]:
pass  # Replace this line with your own solution before checking the model answer.


## Exercise 3.14.10 — Investigate a claim about unresolved cases

A service manager says that the service type with the largest number of received cases must also have the largest unresolved workload. The statement sounds reasonable, but it combines totals, differences and percentages as if they were the same measure. The raw file contains records where resolved cases exceed received cases and one record where received cases are negative. Those values can produce impossible unresolved counts or rates above one hundred percent. A row-level rate also answers a different question from a rate calculated from grouped totals. The service labels contain several variants that may fragment groups before any comparison is made. You are asked to test the manager's statement without silently correcting or deleting the suspicious records. Your answer must show the formulas, flag invalid inputs and explain which conclusion remains preliminary.

### Your task

1. Load and verify the raw dataset in a fresh block of code.
2. Create `unresolved_cases` and a row-level `resolution_rate_pct`, using only rows with a positive denominator for the rate.
3. Display records with negative received cases, negative unresolved cases or rates above 100%.
4. Calculate service-type totals for received and resolved cases, then derive unresolved totals and an aggregate resolution rate for every raw service label.
5. Compare the ranking suggested by received totals with the ranking suggested by unresolved totals.
6. Explain why the raw comparison cannot yet settle the manager's claim.


In [ ]:
pass  # Replace this line with your own solution before checking the model answer.


## Exercise 3.14.11 — Check whether service-channel counts agree

A digital-inclusion group wants to know how much of the recorded service activity happened digitally rather than in person. The dataset contains `digital_cases`, `in_person_cases` and the wider `cases_received` total for every record. Someone proposes dividing digital cases by received cases immediately and presenting the result as digital uptake. Before doing that, the group must check whether the two channel counts actually add up to the stated total. A zero or negative denominator would also make a percentage meaningless or misleading. Differences across service types may reflect the composition of the table rather than the behaviour of individual residents. The file is synthetic and its raw category variants can split otherwise related records. Your task is to create an auditable channel comparison and state exactly what its denominator represents.

### Your task

1. Load the raw data, confirm the columns needed for the calculation and inspect a reproducible sample.
2. Create `recorded_channel_total`, `channel_gap` and `digital_share_pct` for rows with positive received cases.
3. Display every row where the channel total differs from `cases_received` or where received cases are not positive.
4. Calculate digital and in-person totals by raw service type and derive a grouped digital share from those totals.
5. Create a crosstab showing service type against whether digital cases exceed in-person cases.
6. Write a cautious interpretation that names the unit of observation, denominator and at least two data limitations.


In [ ]:
pass  # Replace this line with your own solution before checking the model answer.


## Exercise 3.14.12 — Build a first-response data-quality register

A data steward has one hour to prepare a first-response note before meeting the person who supplied the CSV file. The steward must not clean the data because the meaning of several unusual values is still unknown. Instead, the note should identify completeness, uniqueness, consistency, validity and cross-field logic concerns. Missing feedback and missing satisfaction need to be counted separately because they affect different analyses. Exact duplicate rows are not the same as identifiers that appear on more than one non-identical row. Numeric-looking and date-like text must be tested safely without overwriting the original columns. Category variants should be listed so the data owner can confirm whether they represent the same places and services. Your final register must connect every detected issue to an affected-row count and a reason for follow-up.

### Your task

1. Load and verify the untouched raw CSV.
2. Count missing values, exact duplicate rows and rows carrying repeated identifiers.
3. List raw city and service labels with their frequencies.
4. Safely test `resolution_days` as numeric and `report_month` as a date without replacing either raw column.
5. Create masks for negative counts, out-of-range satisfaction and resolved cases greater than received cases.
6. Assemble a compact issue register with issue name, affected-row count, detection rule and follow-up reason, then identify which issues need the data owner's decision.


In [ ]:
pass  # Replace this line with your own solution before checking the model answer.


## Exercise 3.14.13 — Examine feedback about waiting and clarity

A service-design workshop wants to discuss whether waiting and unclear information appear in the written feedback. The dataset contains short comments rather than interview transcripts, and three records have no stored feedback. Workshop participants first need to read a sample so that the coding rules are connected to actual language. A search for `wait` may miss phrases such as `took too long`, while a search for `clear` may also match a negated statement. The number of keyword matches is therefore a count produced by a rule, not a complete qualitative finding. The team also wants to compare matched comments with service type and satisfaction, but missing scores change the denominator. No association in this table can demonstrate that waiting or clarity caused a rating. Your task is to make the coding rules transparent, inspect their matches and report their limitations beside the results.

### Your task

1. Load the raw dataset and display a reproducible sample of non-missing feedback.
2. Calculate character and word counts for the non-missing comments and summarise both measures.
3. Create separate Boolean indicators for waiting/delay language and clarity/helpfulness language.
4. Count the indicators, display all matched comments and manually identify at least one possible false positive or false negative.
5. Compare each indicator with service type using crosstabs and with satisfaction using count, mean and median.
6. State the number of matched comments and the smaller or equal number contributing a satisfaction score, then write a non-causal interpretation.


In [ ]:
pass  # Replace this line with your own solution before checking the model answer.


## Exercise 3.14.14 — Test a strong management claim before repeating it

A fictional management slide claims that Copenhagen Transport has the heaviest workload, the poorest satisfaction and the most waiting complaints in the dataset. It then says that long waits caused the low satisfaction scores. You have been asked to check the slide before it is shown in class. The claim combines an exact city-service subset, comparisons with other groups, a simple text rule and a causal conclusion. Raw category variants may exclude relevant records when exact labels are used. Missing and out-of-range satisfaction values may change which records contribute to an average. Keyword matches may describe an association but cannot establish why a score was given. Your review must separate what the code observes, what can be cautiously interpreted and what the dataset cannot support.

### Your task

1. Load and verify the raw file, then inspect city and service-type label frequencies.
2. Select exact `Copenhagen` and `Transport` rows and report their row count, received-case total, mean and median satisfaction, and waiting-rule match count.
3. Create a city-by-service summary so the claimed workload and satisfaction rankings can be compared with other exact-label groups.
4. Check missing or out-of-range satisfaction, impossible case relationships and duplicates within the focal subset and in the complete dataset.
5. Display the focal waiting-related comments and compare their satisfaction denominator with the number of text matches.
6. Write three labelled statements: a direct observation, a cautious interpretation and an unsupported claim. End with a decision to retain, revise or remove the original slide statement.


In [ ]:
pass  # Replace this line with your own solution before checking the model answer.


# STOP — Solutions begin below

The remaining sections are for checking work after an independent attempt. Each coding answer is fully commented and has been executed against the supplied raw CSV.


## Solution 3.1.5 — Describe a dataset before coding

One row could represent one measured organisation in one reporting period, but the unit must be confirmed from documentation. `employees` and `energy_use_kwh` look numeric, `organisation_id` is an identifier, and `comment` is free text. Interpreting energy use requires the period, coverage, unit definition, collection method and treatment of missing values. Several rows may belong to the same organisation, so rows are not automatically organisations.


## Solution 3.2.4 — Find the structural problem

The file appears semicolon-delimited, while `read_csv()` assumes commas by default. Inspect the raw text and source documentation, then test `sep=";"`; do not change the delimiter blindly because a one-column result can also indicate malformed quoting or the wrong file.


## Solution 3.2.5 — Missing does not mean zero

Zero is an observed numeric value. A blank field means no recognised value was stored and may mean missing, not measured or not applicable. Replacing every blank with zero would invent a measurement and can distort totals, averages and quality assessments.


## Solution 3.3.14 — Load and verify

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [35]:
import pandas as pd  # Loads the library feature used by this step.
url = "https://raw.githubusercontent.com/asmrabbi/E26_TAN7_Scripting_CPH/main/data/E26_TAN7_service_experience_raw.csv"  # Stores the calculated or supplied value for later use.
df = pd.read_csv(url)  # Stores the calculated or supplied value for later use.
display(df.head())  # Displays the result so it can be inspected.
print(type(df))  # Displays the result so it can be inspected.


     record_id  ...                                           feedback
0  SR2026-0001  ...  The guidance was clear, but the supporting-doc...
1  SR2026-0002  ...  Several users asked for clearer information ab...
2  SR2026-0003  ...  Useful advice and a friendly response from staff.
3  SR2026-0004  ...  Helpful guidance, but some users reported dela...
4  SR2026-0005  ...  The response was fast and the information was ...

[5 rows x 14 columns]
<class 'pandas.core.frame.DataFrame'>


## Solution 3.3.15 — Explain the file/DataFrame distinction

`df` is an in-memory object created from the downloaded file. Changing the object does not rewrite the remote GitHub resource. A permanent change requires an explicit save operation and, for GitHub, permission plus a commit or upload.


## Solution 3.4.16 — Produce an inspection note

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [36]:
rows, columns = df.shape  # Stores the calculated or supplied value for later use.
print("Rows:", rows)  # Displays the result so it can be inspected.
print("Columns:", columns)  # Displays the result so it can be inspected.
print("Column names:", df.columns.tolist())  # Displays the result so it can be inspected.
display(df.head(3))  # Displays the result so it can be inspected.
display(df.tail(3))  # Displays the result so it can be inspected.
print(df.dtypes)  # Displays the result so it can be inspected.
df.info()  # Runs this part of the documented analysis.


Rows: 121
Columns: 14
Column names: ['record_id', 'report_month', 'city', 'service_type', 'cases_received', 'cases_resolved', 'resolution_days', 'satisfaction_score', 'digital_cases', 'in_person_cases', 'repeat_contacts', 'complaints_received', 'staff_hours', 'feedback']
     record_id  ...                                           feedback
0  SR2026-0001  ...  The guidance was clear, but the supporting-doc...
1  SR2026-0002  ...  Several users asked for clearer information ab...
2  SR2026-0003  ...  Useful advice and a friendly response from staff.

[3 rows x 14 columns]
       record_id  ...                                           feedback
118  SR2026-0119  ...  Collection information was easy to understand ...
119  SR2026-0120  ...  The response was fast and the information was ...
120  SR2026-0027  ...  The digital form worked well, although some in...

[3 rows x 14 columns]
record_id               object
report_month            object
city                    object
service_type 

## Solution 3.4.17 — Why inspect beyond the first five rows?

The first five rows may not contain rare errors or later category variants, and files are often ordered by time or source. Inspect the tail, reproducible samples, structure, missing counts, dtypes, ranges and targeted quality conditions as well.


## Solution 3.5.21 — Column selection

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [37]:
selected = df[["record_id", "service_type", "satisfaction_score", "feedback"]]  # Stores the calculated or supplied value for later use.
display(selected.head())  # Displays the result so it can be inspected.
print(type(selected))  # Displays the result so it can be inspected.


     record_id  ...                                           feedback
0  SR2026-0001  ...  The guidance was clear, but the supporting-doc...
1  SR2026-0002  ...  Several users asked for clearer information ab...
2  SR2026-0003  ...  Useful advice and a friendly response from staff.
3  SR2026-0004  ...  Helpful guidance, but some users reported dela...
4  SR2026-0005  ...  The response was fast and the information was ...

[5 rows x 4 columns]
<class 'pandas.core.frame.DataFrame'>


## Solution 3.5.22 — Filter a numeric condition

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [38]:
at_least_100 = df[df["cases_received"] >= 100]  # Stores the calculated or supplied value for later use.
display(at_least_100)  # Displays the result so it can be inspected.
print("Matching rows:", len(at_least_100))  # Displays the result so it can be inspected.


      record_id report_month        city      service_type  cases_received  cases_resolved resolution_days  satisfaction_score  digital_cases  in_person_cases  repeat_contacts  complaints_received  staff_hours                                                                            feedback
0   SR2026-0001   2026-01-01  Copenhagen           Housing             130             116             8.0                 3.5             88               42               17                   10         87.8     The guidance was clear, but the supporting-document requirements caused delays.
1   SR2026-0002   2026-01-01  Copenhagen         Transport             124             110             5.1                 3.9             92               32               10                    8         54.7                  Several users asked for clearer information about service changes.
2   SR2026-0003   2026-01-01  Copenhagen        Employment             105              93             5.5            

## Solution 3.5.23 — Combine a category and numeric condition

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [39]:
transport_over_90 = df[(df["service_type"] == "Transport") & (df["cases_received"] > 90)]  # Stores the calculated or supplied value for later use.
display(transport_over_90[["record_id", "city", "cases_received"]])  # Displays the result so it can be inspected.


       record_id          city  cases_received
1    SR2026-0002    Copenhagen             124
6    SR2026-0007   copenhagen               95
21   SR2026-0022    Copenhagen             126
26   SR2026-0027        Aarhus             110
41   SR2026-0042    Copenhagen             132
46   SR2026-0047        Aarhus             119
51   SR2026-0052        Odense              93
61   SR2026-0062    Copenhagen             129
66   SR2026-0067        Aarhus             102
81   SR2026-0082        aarhus             144
86   SR2026-0087        Aarhus             122
91   SR2026-0092        Odense              92
101  SR2026-0102    Copenhagen             123
106  SR2026-0107        Aarhus             102
120  SR2026-0027        Aarhus             110


## Solution 3.5.24 — Housing reports for review

Run this solution after attempting the exercise. Preserve the raw values and explain the result in ordinary language.


### Housing selection


In [48]:
mask = (df["service_type"] == "Housing") & (  # Require Housing and a second condition.
    df["cases_received"] > 120  # Also require more than 120 received cases.
)  # Finish the rule.
selected = df.loc[mask, ["city", "cases_received"]]  # Keep matching rows and two columns.
print("Matching rows:", len(selected))  # Count all matching reports.
display(selected.head(3))  # Show only the first three.


Matching rows: 10
          city  cases_received
0   Copenhagen             130
20  Copenhagen             157
25      Aarhus             136


## Solution 3.6.4 — Classify the course variables

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [40]:
roles = {"record_id": "identifier", "report_month": "date-like", "city": "categorical", "service_type": "categorical", "cases_received": "numeric count", "cases_resolved": "numeric count", "resolution_days": "numeric-intended measure", "satisfaction_score": "numeric rating", "digital_cases": "numeric count", "in_person_cases": "numeric count", "repeat_contacts": "numeric count", "complaints_received": "numeric count", "staff_hours": "numeric measure", "feedback": "free text"}  # Stores the calculated or supplied value for later use.
role_table = pd.DataFrame({"variable": list(roles), "conceptual_type": list(roles.values())})  # Stores the calculated or supplied value for later use.
role_table["pandas_dtype"] = role_table["variable"].map(df.dtypes.astype(str))  # Stores the calculated or supplied value for later use.
display(role_table)  # Displays the result so it can be inspected.


               variable           conceptual_type pandas_dtype
0             record_id                identifier       object
1          report_month                 date-like       object
2                  city               categorical       object
3          service_type               categorical       object
4        cases_received             numeric count        int64
5        cases_resolved             numeric count        int64
6       resolution_days  numeric-intended measure       object
7    satisfaction_score            numeric rating      float64
8         digital_cases             numeric count        int64
9       in_person_cases             numeric count        int64
10      repeat_contacts             numeric count        int64
11  complaints_received             numeric count        int64
12          staff_hours           numeric measure      float64
13             feedback                 free text       object


## Solution 3.6.5 — Explain why this statement is wrong

Technical storage does not determine analytical meaning. `record_id` labels records; an arithmetic average of identifiers has no substantive interpretation even if pandas stores the labels in a numeric dtype.


## Solution 3.7.15 — Produce a numeric summary

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [41]:
resolved = df["cases_resolved"]  # Stores the calculated or supplied value for later use.
print("Count:", resolved.count())  # Displays the result so it can be inspected.
print("Sum:", resolved.sum())  # Displays the result so it can be inspected.
print("Mean:", resolved.mean())  # Displays the result so it can be inspected.
print("Median:", resolved.median())  # Displays the result so it can be inspected.
print("Minimum:", resolved.min())  # Displays the result so it can be inspected.
print("Maximum:", resolved.max())  # Displays the result so it can be inspected.
print("Standard deviation:", resolved.std())  # Displays the result so it can be inspected.
print(resolved.describe())  # Displays the result so it can be inspected.


Count: 121
Sum: 11958
Mean: 98.82644628099173
Median: 95.0
Minimum: 0
Maximum: 176
Standard deviation: 26.195953149914338
count    121.000000
mean      98.826446
std       26.195953
min        0.000000
25%       82.000000
50%       95.000000
75%      113.000000
max      176.000000
Name: cases_resolved, dtype: float64


## Solution 3.7.16 — Mean or median?

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [42]:
mean_received = df["cases_received"].mean()  # Stores the calculated or supplied value for later use.
median_received = df["cases_received"].median()  # Stores the calculated or supplied value for later use.
print("Mean:", round(mean_received, 2))  # Displays the result so it can be inspected.
print("Median:", median_received)  # Displays the result so it can be inspected.
display(df[df["cases_received"] < 0])  # Displays the result so it can be inspected.


Mean: 104.57
Median: 102.0
      record_id  ...                                           feedback
53  SR2026-0054  ...  Helpful guidance, but some users reported dela...

[1 rows x 14 columns]


## Solution 3.7.17 — Count the denominator explicitly

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [43]:
mean_score = df["satisfaction_score"].mean()  # Stores the calculated or supplied value for later use.
valid_scores = df["satisfaction_score"].count()  # Stores the calculated or supplied value for later use.
print("Mean satisfaction:", round(mean_score, 2))  # Displays the result so it can be inspected.
print("Scores included:", valid_scores)  # Displays the result so it can be inspected.


Mean satisfaction: 4.11
Scores included: 117


## Solution 3.8.10 — Derive a difference

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [44]:
df["unresolved_cases"] = df["cases_received"] - df["cases_resolved"]  # Stores the calculated or supplied value for later use.
print(df["unresolved_cases"].nsmallest(5))  # Displays the result so it can be inspected.
display(df[df["unresolved_cases"] < 0][["record_id", "cases_received", "cases_resolved", "unresolved_cases"]])  # Displays the result so it can be inspected.


11   -7
53   -4
89   -3
79    1
91    1
Name: unresolved_cases, dtype: int64
      record_id  cases_received  cases_resolved  unresolved_cases
11  SR2026-0012              80              87                -7
53  SR2026-0054              -4               0                -4
89  SR2026-0090             159             162                -3


## Solution 3.8.11 — Explain two rate formulas

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [45]:
row_rates = df["cases_resolved"] / df["cases_received"] * 100  # Stores the calculated or supplied value for later use.
mean_row_rate = row_rates.mean()  # Stores the calculated or supplied value for later use.
overall_rate = df["cases_resolved"].sum() / df["cases_received"].sum() * 100  # Stores the calculated or supplied value for later use.
print("Mean row rate:", round(mean_row_rate, 2))  # Displays the result so it can be inspected.
print("Aggregate rate:", round(overall_rate, 2))  # Displays the result so it can be inspected.


Mean row rate: 93.66
Aggregate rate: 94.51


## Solution 3.8.12 — Calculate a missing percentage

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [46]:
missing_count = df["satisfaction_score"].isna().sum()  # Stores the calculated or supplied value for later use.
missing_percentage = missing_count / len(df) * 100  # Stores the calculated or supplied value for later use.
print("Missing satisfaction values:", missing_count)  # Displays the result so it can be inspected.
print("Missing percentage:", round(missing_percentage, 2))  # Displays the result so it can be inspected.


Missing satisfaction values: 4
Missing percentage: 3.31


## Solution 3.9.14 — Frequencies and percentages

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [47]:
print(df["service_type"].unique())  # Displays the result so it can be inspected.
print("Distinct raw labels:", df["service_type"].nunique())  # Displays the result so it can be inspected.
print(df["service_type"].value_counts())  # Displays the result so it can be inspected.
print(df["service_type"].value_counts(normalize=True).mul(100).round(2))  # Displays the result so it can be inspected.
print("Modes:", df["service_type"].mode().tolist())  # Displays the result so it can be inspected.


['Housing' 'Transport' 'Employment' 'Waste' 'Citizen Services'
 'Transport ' 'Citizen services' 'WASTE' 'Employment Services']
Distinct raw labels: 9
service_type
Transport              24
Employment             24
Waste                  24
Citizen Services       24
Housing                21
Transport               1
Citizen services        1
WASTE                   1
Employment Services     1
Name: count, dtype: int64
service_type
Transport              19.83
Employment             19.83
Waste                  19.83
Citizen Services       19.83
Housing                17.36
Transport               0.83
Citizen services        0.83
WASTE                   0.83
Employment Services     0.83
Name: proportion, dtype: float64
Modes: ['Citizen Services', 'Employment', 'Transport', 'Waste']


## Solution 3.9.15 — Grouped summary

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [48]:
service_summary = df.groupby("service_type")["cases_received"].agg(["count", "mean", "median", "min", "max"])  # Stores the calculated or supplied value for later use.
display(service_summary.round(2))  # Displays the result so it can be inspected.


                     count    mean  median  min  max
service_type                                        
Citizen Services        24  129.92   122.0   94  185
Citizen services         1   99.00    99.0   99   99
Employment              24   88.04    86.5   62  129
Employment Services      1   97.00    97.0   97   97
Housing                 21  115.86   113.0   86  157
Transport               24  102.54    98.5   75  144
Transport                1  107.00   107.0  107  107
WASTE                    1   90.00    90.0   90   90
Waste                   24   88.96    89.5   -4  138


## Solution 3.9.16 — Discover category fragmentation

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [49]:
city_summary = df.groupby("city")["cases_received"].agg(["count", "sum", "mean"])  # Stores the calculated or supplied value for later use.
display(city_summary.round(2))  # Displays the result so it can be inspected.
print("Raw city labels:", df["city"].unique().tolist())  # Displays the result so it can be inspected.


              count   sum    mean
city                             
 copenhagen       1    95   95.00
AALBORG           1    76   76.00
Aalborg          28  2386   85.21
Aarhus           30  3349  111.63
Copenhagen       29  3807  131.28
Koebenhavn        1    76   76.00
Odense           29  2642   91.10
Odense            1    78   78.00
aarhus            1   144  144.00
Raw city labels: ['Copenhagen', 'Aarhus', ' copenhagen ', 'Odense', 'Aalborg', 'AALBORG', 'Odense ', 'aarhus', 'Koebenhavn']


## Solution 3.9.17 — Create a crosstab

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [50]:
city_service = pd.crosstab(df["city"], df["service_type"])  # Stores the calculated or supplied value for later use.
display(city_service)  # Displays the result so it can be inspected.
row_percent = pd.crosstab(df["city"], df["service_type"], normalize="index").mul(100)  # Stores the calculated or supplied value for later use.
display(row_percent.round(2))  # Displays the result so it can be inspected.


service_type  Citizen Services  Citizen services  ...  WASTE  Waste
city                                              ...              
 copenhagen                  0                 0  ...      0      0
AALBORG                      0                 0  ...      0      0
Aalborg                      6                 1  ...      0      5
Aarhus                       6                 0  ...      0      6
Copenhagen                   6                 0  ...      0      6
Koebenhavn                   0                 0  ...      0      1
Odense                       6                 0  ...      1      6
Odense                       0                 0  ...      0      0
aarhus                       0                 0  ...      0      0

[9 rows x 9 columns]
service_type  Citizen Services  Citizen services  ...  WASTE   Waste
city                                              ...               
 copenhagen               0.00              0.00  ...   0.00    0.00
AALBORG                

## Solution 3.10.19 — Build a quality audit

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [51]:
print("Missing counts")  # Displays the result so it can be inspected.
print(df.isna().sum())  # Displays the result so it can be inspected.
print("Missing percentages")  # Displays the result so it can be inspected.
print(df.isna().mean().mul(100).round(2))  # Displays the result so it can be inspected.
print("Exact duplicates:", df.duplicated().sum())  # Displays the result so it can be inspected.
display(df[df.duplicated(keep=False)])  # Displays the result so it can be inspected.
display(df[df["record_id"].duplicated(keep=False)])  # Displays the result so it can be inspected.
print("Raw cities:", df["city"].unique().tolist())  # Displays the result so it can be inspected.
display(df[df["cases_received"] < 0])  # Displays the result so it can be inspected.
sat_problem = (df["satisfaction_score"] < 1) | (df["satisfaction_score"] > 5)  # Stores the calculated or supplied value for later use.
display(df[sat_problem])  # Displays the result so it can be inspected.
display(df[df["cases_resolved"] > df["cases_received"]])  # Displays the result so it can be inspected.
resolution_preview = pd.to_numeric(df["resolution_days"], errors="coerce")  # Stores the calculated or supplied value for later use.
display(df.loc[df["resolution_days"].notna() & resolution_preview.isna(), ["record_id", "resolution_days"]])  # Displays the result so it can be inspected.
month_preview = pd.to_datetime(df["report_month"], errors="coerce")  # Stores the calculated or supplied value for later use.
display(df.loc[month_preview.isna(), ["record_id", "report_month"]])  # Displays the result so it can be inspected.


Missing counts
record_id              0
report_month           0
city                   0
service_type           0
cases_received         0
cases_resolved         0
resolution_days        0
satisfaction_score     4
digital_cases          0
in_person_cases        0
repeat_contacts        0
complaints_received    0
staff_hours            0
feedback               3
unresolved_cases       0
dtype: int64
Missing percentages
record_id              0.00
report_month           0.00
city                   0.00
service_type           0.00
cases_received         0.00
cases_resolved         0.00
resolution_days        0.00
satisfaction_score     3.31
digital_cases          0.00
in_person_cases        0.00
repeat_contacts        0.00
complaints_received    0.00
staff_hours            0.00
feedback               2.48
unresolved_cases       0.00
dtype: float64
Exact duplicates: 1
       record_id  ... unresolved_cases
26   SR2026-0027  ...                8
120  SR2026-0027  ...                8

[2 r

## Solution 3.10.20 — Classify the quality issue

1 completeness; 2 uniqueness; 3 consistency; 4 validity; 5 validity; 6 logical consistency. Plausibility can overlap with several items, but the documented rule should be named.


## Solution 3.10.21 — checking a satisfaction claim

Run this solution after attempting the exercise. Preserve the raw values and explain the result in ordinary language.


### the quality checks


In [61]:
scores = df["satisfaction_score"]  # Select the recorded score column.
outside = (scores < 1) | (scores > 5)  # Flag scores below 1 or above 5.
print("Missing scores:", scores.isna().sum())  # Count recognised missing scores separately.
columns = ["record_id", "satisfaction_score"]  # Choose the identifiers and stored scores.
display(df.loc[outside, columns])  # Display the out-of-range records.


Missing scores: 4
      record_id  satisfaction_score
38  SR2026-0039                 6.2
72  SR2026-0073                 0.0


### a cautious finding


In [62]:
scores = df["satisfaction_score"]  # Select the source scores again.
print("Recorded scores:", scores.count())  # Count the values that contribute to the mean.
print("Raw mean:", round(scores.mean(), 2))  # Calculate and display the raw mean.


Recorded scores: 117
Raw mean: 4.11


**Model interpretation**

The raw mean is 4.11 across 117 recorded scores. Four scores are missing and two fall outside the expected range, so the average needs review.


## Solution 3.11.14 — Explore text before coding themes

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [52]:
feedback = df["feedback"].dropna()  # Stores the calculated or supplied value for later use.
print(feedback.to_string(index=False))  # Displays the result so it can be inspected.
print("Available comments:", feedback.count())  # Displays the result so it can be inspected.


The guidance was clear, but the supporting-docu...
Several users asked for clearer information abo...
 Useful advice and a friendly response from staff.
Helpful guidance, but some users reported delay...
The response was fast and the information was u...
Helpful staff explained the housing process, al...
Several users asked for clearer information abo...
The answer was useful, although the waiting tim...
Collection information was easy to understand a...
                Friendly staff and a clear answer.
The online application was useful, but several ...
The response was quick, but route-change inform...
 Useful advice and a friendly response from staff.
Helpful guidance, but some users reported delay...
The response was fast and the information was u...
The guidance was clear, but the supporting-docu...
The digital form worked well, although some ins...
The issue was resolved quickly after the first ...
The digital form was convenient and staff expla...
Helpful staff explained the hou

## Solution 3.11.15 — Search and validate a keyword rule

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [53]:
speed_mask = df["feedback"].str.contains("quick|fast", case=False, na=False, regex=True)  # Stores the calculated or supplied value for later use.
print("Speed-rule matches:", int(speed_mask.sum()))  # Displays the result so it can be inspected.
print(df.loc[speed_mask, "feedback"].to_string(index=False))  # Displays the result so it can be inspected.


Speed-rule matches: 32
The response was fast and the information was u...
Collection information was easy to understand a...
The response was quick, but route-change inform...
The response was fast and the information was u...
The issue was resolved quickly after the first ...
Quick and helpful service with clear instructions.
Collection information was easy to understand a...
                Fast response and useful guidance.
                Fast response and useful guidance.
The issue was resolved quickly after the first ...
The response was quick, but route-change inform...
The issue was resolved quickly after the first ...
The response was fast and the information was u...
The response was quick, but route-change inform...
The issue was resolved quickly after the first ...
The response was fast and the information was u...
The response was quick, but route-change inform...
The issue was resolved quickly after the first ...
The response was fast and the information was u...
The resp

## Solution 3.11.16 — Compare quantitative and qualitative evidence

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [54]:
wait_mask = df["feedback"].str.contains("wait|long", case=False, na=False, regex=True)  # Stores the calculated or supplied value for later use.
wait_rows = df.loc[wait_mask, ["record_id", "service_type", "satisfaction_score", "feedback"]]  # Stores the calculated or supplied value for later use.
display(wait_rows)  # Displays the result so it can be inspected.
print("Matched comments:", len(wait_rows))  # Displays the result so it can be inspected.
print("Matched comments with scores:", wait_rows["satisfaction_score"].notna().sum())  # Displays the result so it can be inspected.


       record_id  ...                                           feedback
5    SR2026-0006  ...  Helpful staff explained the housing process, a...
7    SR2026-0008  ...  The answer was useful, although the waiting ti...
20   SR2026-0021  ...  Helpful staff explained the housing process, a...
22   SR2026-0023  ...  The answer was useful, although the waiting ti...
25   SR2026-0026  ...  Helpful staff explained the housing process, a...
30   SR2026-0031  ...  The wait was not long, and the final response ...
37   SR2026-0038  ...  The answer was useful, although the waiting ti...
55   SR2026-0056  ...  Helpful staff explained the housing process, a...
57   SR2026-0058  ...  The answer was useful, although the waiting ti...
72   SR2026-0073  ...  The answer was useful, although the waiting ti...
110  SR2026-0111  ...  Helpful staff explained the housing process, a...

[11 rows x 4 columns]
Matched comments: 11
Matched comments with scores: 10


## Solution 3.12.9 — Design a coding rule

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [55]:
df["mentions_clarity"] = df["feedback"].str.contains("clear", case=False, na=False)  # Stores the calculated or supplied value for later use.
print("Clarity matches:", int(df["mentions_clarity"].sum()))  # Displays the result so it can be inspected.
print(df.loc[df["mentions_clarity"], "feedback"].to_string(index=False))  # Displays the result so it can be inspected.
print("Three non-matches for manual review")  # Displays the result so it can be inspected.
print(df.loc[~df["mentions_clarity"] & df["feedback"].notna(), "feedback"].head(3).to_string(index=False))  # Displays the result so it can be inspected.


Clarity matches: 40
The guidance was clear, but the supporting-docu...
Several users asked for clearer information abo...
Several users asked for clearer information abo...
                Friendly staff and a clear answer.
The guidance was clear, but the supporting-docu...
The digital form was convenient and staff expla...
Clear information and a helpful answer about th...
Quick and helpful service with clear instructions.
The guidance was clear but the process required...
The digital form was convenient and staff expla...
The wait was not long, and the final response w...
The digital form was convenient and staff expla...
The digital form was convenient and staff expla...
Several users asked for clearer information abo...
The answer solved the problem, but repeated con...
Clear information and a helpful answer about th...
The digital form was convenient and staff expla...
The guidance was clear, but the supporting-docu...
The guidance was clear, but the supporting-docu...
Quick and h

## Solution 3.12.10 — Compare code and score

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [56]:
comparison = df.groupby("mentions_clarity")["satisfaction_score"].agg(["count", "mean", "median"])  # Stores the calculated or supplied value for later use.
display(comparison.round(2))  # Displays the result so it can be inspected.


                  count  mean  median
mentions_clarity                     
False                78  4.08     4.1
True                 39  4.17     4.2


## Solution 3.12.11 — Multiple codes per comment

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [57]:
df["mentions_speed"] = df["feedback"].str.contains("quick|fast", case=False, na=False, regex=True)  # Stores the calculated or supplied value for later use.
df["mentions_helpful"] = df["feedback"].str.contains("helpful", case=False, na=False)  # Stores the calculated or supplied value for later use.
both_mask = df["mentions_speed"] & df["mentions_helpful"]  # Stores the calculated or supplied value for later use.
display(df.loc[both_mask, ["record_id", "feedback"]])  # Displays the result so it can be inspected.
print("Comments with both codes:", int(both_mask.sum()))  # Displays the result so it can be inspected.


      record_id                                           feedback
24  SR2026-0025  Quick and helpful service with clear instructi...
69  SR2026-0070  Quick and helpful service with clear instructi...
99  SR2026-0100  Quick and helpful service with clear instructi...
Comments with both codes: 3


## Solution 3.13.3 — Create a data dictionary

The code below gives one reproducible solution. Describe the result in relation to the unit of observation, the denominator and any visible raw-data limitation.


In [58]:
dictionary_rows = [{"variable": column, "pandas_dtype": str(df[column].dtype), "meaning_or_rule": "Complete from the published Lecture 5 data dictionary"} for column in df.columns]  # Stores the calculated or supplied value for later use.
data_dictionary = pd.DataFrame(dictionary_rows)  # Stores the calculated or supplied value for later use.
display(data_dictionary)  # Displays the result so it can be inspected.


               variable pandas_dtype                                        meaning_or_rule
0             record_id       object  Complete from the published Lecture 5 data dictionary
1          report_month       object  Complete from the published Lecture 5 data dictionary
2                  city       object  Complete from the published Lecture 5 data dictionary
3          service_type       object  Complete from the published Lecture 5 data dictionary
4        cases_received        int64  Complete from the published Lecture 5 data dictionary
5        cases_resolved        int64  Complete from the published Lecture 5 data dictionary
6       resolution_days       object  Complete from the published Lecture 5 data dictionary
7    satisfaction_score      float64  Complete from the published Lecture 5 data dictionary
8         digital_cases        int64  Complete from the published Lecture 5 data dictionary
9       in_person_cases        int64  Complete from the published Lecture 5 data

## Solution 3.13.4 — Rewrite an overclaim

In the raw synthetic teaching table, the exact `Transport` label has 2,461 received cases across 24 rows. This does not establish that public transport is Denmark's biggest problem because the data are synthetic, the unit is a service-category-city-month record, categories are inconsistent and the file is not representative of residents or cities.


## Solution 3.13.5 — Separate three levels of statement

Observation: four exact service labels each occur in 24 of 121 raw rows. Cautious interpretation: the raw table is nearly balanced across several intended service categories, though inconsistent variants split some groups. Unsupported claim: real residents use those services equally often.


# Tutorial 3.14 integrated exercise solutions

Each solution is deliberately standalone. It imports pandas, defines the exact raw-data URL and loads the CSV again, so a teacher can run any single solution from a fresh Colab runtime.


## Solution 3.14.9 — Prepare a two-city service briefing

Run this complete block, inspect the displayed rows and summaries, and compare the final interpretation with your own wording.


In [59]:
import pandas as pd  # Imports pandas so this solution can run in a fresh Colab cell.

data_url = "https://raw.githubusercontent.com/asmrabbi/E26_TAN7_Scripting_CPH/main/data/E26_TAN7_service_experience_raw.csv"  # Stores the versioned raw-data address.
df = pd.read_csv(data_url)  # Loads a new DataFrame for this exercise only.
assert df.shape == (121, 14), "The wrong dataset version was loaded."

print("Dataset shape:", df.shape)
display(df.head())

print("\nEvery raw city label")
print(df["city"].value_counts(dropna=False))

selected_cities = ["Copenhagen", "Aalborg"]
city_records = df[df["city"].isin(selected_cities)].copy()

briefing_columns = [
    "record_id", "report_month", "city", "service_type",
    "cases_received", "cases_resolved", "satisfaction_score", "feedback"
]
display(city_records[briefing_columns].head(10))

city_summary = city_records.groupby("city").agg(
    row_count=("record_id", "count"),
    cases_received_total=("cases_received", "sum"),
    cases_received_mean=("cases_received", "mean"),
    cases_resolved_total=("cases_resolved", "sum"),
    satisfaction_n=("satisfaction_score", "count"),
    satisfaction_mean=("satisfaction_score", "mean"),
    satisfaction_median=("satisfaction_score", "median")
)
display(city_summary.round(2))

city_records["mentions_wait"] = city_records["feedback"].str.contains(
    "wait|long|delay", case=False, na=False, regex=True
)
print("\nWaiting-rule matches by city")
print(pd.crosstab(city_records["city"], city_records["mentions_wait"]))
display(city_records.loc[city_records["mentions_wait"], ["record_id", "city", "feedback"]])

print("\nInterpretation")
print("These exact-label results describe synthetic city-service-month records, not residents.")
print("Raw city variants are excluded from the two exact-label groups and need owner review.")
print("Satisfaction counts are the denominators for satisfaction means; keyword matches describe only the stated rule.")

Dataset shape: (121, 14)
     record_id report_month        city      service_type  cases_received  cases_resolved resolution_days  satisfaction_score  digital_cases  in_person_cases  repeat_contacts  complaints_received  staff_hours                                                                         feedback
0  SR2026-0001   2026-01-01  Copenhagen           Housing             130             116             8.0                 3.5             88               42               17                   10         87.8  The guidance was clear, but the supporting-document requirements caused delays.
1  SR2026-0002   2026-01-01  Copenhagen         Transport             124             110             5.1                 3.9             92               32               10                    8         54.7               Several users asked for clearer information about service changes.
2  SR2026-0003   2026-01-01  Copenhagen        Employment             105              93             5.5

## Solution 3.14.10 — Investigate a claim about unresolved cases

Run this complete block, inspect the displayed rows and summaries, and compare the final interpretation with your own wording.


In [60]:
import pandas as pd  # Imports pandas so this solution is independent of earlier cells.

data_url = "https://raw.githubusercontent.com/asmrabbi/E26_TAN7_Scripting_CPH/main/data/E26_TAN7_service_experience_raw.csv"  # Identifies the unchanged raw teaching file.
df = pd.read_csv(data_url)  # Loads the complete dataset again.
assert df.shape == (121, 14), "The wrong dataset version was loaded."

df["unresolved_cases"] = df["cases_received"] - df["cases_resolved"]
positive_received = df["cases_received"] > 0
df.loc[positive_received, "resolution_rate_pct"] = (
    df.loc[positive_received, "cases_resolved"]
    / df.loc[positive_received, "cases_received"]
    * 100
)

problem_mask = (
    (df["cases_received"] < 0)
    | (df["unresolved_cases"] < 0)
    | (df["resolution_rate_pct"] > 100)
)
problem_columns = [
    "record_id", "service_type", "cases_received", "cases_resolved",
    "unresolved_cases", "resolution_rate_pct"
]
print("Rows requiring review:", int(problem_mask.sum()))
display(df.loc[problem_mask, problem_columns])

service_totals = df.groupby("service_type")[["cases_received", "cases_resolved"]].sum()
service_totals["unresolved_total"] = (
    service_totals["cases_received"] - service_totals["cases_resolved"]
)
service_totals["aggregate_resolution_rate_pct"] = (
    service_totals["cases_resolved"] / service_totals["cases_received"] * 100
)

print("\nGrouped results in received-total order")
display(service_totals.sort_values("cases_received", ascending=False).round(2))
print("\nGrouped results in unresolved-total order")
display(service_totals.sort_values("unresolved_total", ascending=False).round(2))

print("\nRaw service labels")
print(df["service_type"].value_counts(dropna=False))
print("\nConclusion")
print("Received volume and unresolved workload are different measures and do not have to rank groups identically.")
print("The raw result remains preliminary because invalid case relationships and fragmented labels affect the grouped totals.")

Rows requiring review: 3
      record_id      service_type  cases_received  cases_resolved  unresolved_cases  resolution_rate_pct
11  SR2026-0012         Transport              80              87                -7           108.750000
53  SR2026-0054             Waste              -4               0                -4                  NaN
89  SR2026-0090  Citizen Services             159             162                -3           101.886792

Grouped results in received-total order
                     cases_received  cases_resolved  unresolved_total  aggregate_resolution_rate_pct
service_type                                                                                        
Citizen Services               3118            3004               114                          96.34
Transport                      2461            2345               116                          95.29
Housing                        2433            2235               198                          91.86
Waste    

## Solution 3.14.11 — Check whether service-channel counts agree

Run this complete block, inspect the displayed rows and summaries, and compare the final interpretation with your own wording.


In [61]:
import pandas as pd  # Imports pandas for a complete standalone solution.

data_url = "https://raw.githubusercontent.com/asmrabbi/E26_TAN7_Scripting_CPH/main/data/E26_TAN7_service_experience_raw.csv"  # Stores the raw CSV location.
df = pd.read_csv(data_url)  # Loads the source data without changing the file.
assert df.shape == (121, 14), "The wrong dataset version was loaded."

required_columns = ["cases_received", "digital_cases", "in_person_cases"]
print("Required columns present:", all(column in df.columns for column in required_columns))
display(df.sample(5, random_state=42)[["record_id", "service_type"] + required_columns])

df["recorded_channel_total"] = df["digital_cases"] + df["in_person_cases"]
df["channel_gap"] = df["cases_received"] - df["recorded_channel_total"]
positive_received = df["cases_received"] > 0
df.loc[positive_received, "digital_share_pct"] = (
    df.loc[positive_received, "digital_cases"]
    / df.loc[positive_received, "cases_received"]
    * 100
)

channel_problem = (df["channel_gap"] != 0) | (~positive_received)
print("Channel or denominator issues:", int(channel_problem.sum()))
display(df.loc[channel_problem, [
    "record_id", "cases_received", "digital_cases", "in_person_cases",
    "recorded_channel_total", "channel_gap", "digital_share_pct"
]])

channel_summary = df.groupby("service_type")[["digital_cases", "in_person_cases"]].sum()
channel_summary["channel_total"] = (
    channel_summary["digital_cases"] + channel_summary["in_person_cases"]
)
channel_summary["grouped_digital_share_pct"] = (
    channel_summary["digital_cases"] / channel_summary["channel_total"] * 100
)
display(channel_summary.round(2))

df["digital_majority"] = df["digital_cases"] > df["in_person_cases"]
display(pd.crosstab(df["service_type"], df["digital_majority"]))

print("\nInterpretation")
print("The denominator is recorded cases in each service-category-city-month row, not people.")
print("Raw category variants fragment groups, and any channel gap must be investigated before reporting uptake.")
print("Because the dataset is synthetic, the percentages cannot describe real service behaviour.")

Required columns present: True
      record_id      service_type  cases_received  digital_cases  in_person_cases
44  SR2026-0045  Citizen Services             185            154               31
47  SR2026-0048        Employment              96             70               26
4   SR2026-0005  Citizen Services             166            137               29
55  SR2026-0056           Housing              93             56               37
26  SR2026-0027         Transport             110             79               31
Channel or denominator issues: 2
      record_id  cases_received  digital_cases  in_person_cases  recorded_channel_total  channel_gap  digital_share_pct
33  SR2026-0034              82             57               43                     100          -18          69.512195
53  SR2026-0054              -4              0                0                       0           -4                NaN
                     digital_cases  in_person_cases  channel_total  grouped_digital_

## Solution 3.14.12 — Build a first-response data-quality register

Run this complete block, inspect the displayed rows and summaries, and compare the final interpretation with your own wording.


In [62]:
import pandas as pd  # Imports pandas for this independent audit.

data_url = "https://raw.githubusercontent.com/asmrabbi/E26_TAN7_Scripting_CPH/main/data/E26_TAN7_service_experience_raw.csv"  # Points to the versioned raw teaching data.
df = pd.read_csv(data_url)  # Loads a fresh untouched DataFrame.
assert df.shape == (121, 14), "The wrong dataset version was loaded."

print("Missing values by column")
print(df.isna().sum())

exact_duplicate_mask = df.duplicated(keep=False)
repeated_id_mask = df["record_id"].duplicated(keep=False)
resolution_preview = pd.to_numeric(df["resolution_days"], errors="coerce")
month_preview = pd.to_datetime(df["report_month"], errors="coerce")

non_numeric_resolution_mask = df["resolution_days"].notna() & resolution_preview.isna()
invalid_month_mask = df["report_month"].notna() & month_preview.isna()
negative_received_mask = df["cases_received"] < 0
negative_complaints_mask = df["complaints_received"] < 0
invalid_satisfaction_mask = (
    (df["satisfaction_score"] < 1) | (df["satisfaction_score"] > 5)
)
resolved_exceeds_received_mask = df["cases_resolved"] > df["cases_received"]

issue_rows = [
    ["Missing satisfaction", int(df["satisfaction_score"].isna().sum()), "isna()", "Changes the score denominator"],
    ["Missing feedback", int(df["feedback"].isna().sum()), "isna()", "Reduces text-analysis coverage"],
    ["Exact duplicate participation", int(exact_duplicate_mask.sum()), "duplicated(keep=False)", "May double-count the same stored row"],
    ["Repeated identifier participation", int(repeated_id_mask.sum()), "record_id duplicated", "May represent reuse or a legitimate revision"],
    ["Non-numeric resolution value", int(non_numeric_resolution_mask.sum()), "to_numeric(errors='coerce')", "Prevents direct numeric analysis"],
    ["Invalid report month", int(invalid_month_mask.sum()), "to_datetime(errors='coerce')", "Prevents valid time ordering"],
    ["Negative received cases", int(negative_received_mask.sum()), "cases_received < 0", "Conflicts with a count interpretation"],
    ["Negative complaints", int(negative_complaints_mask.sum()), "complaints_received < 0", "Conflicts with a count interpretation"],
    ["Satisfaction outside 1–5", int(invalid_satisfaction_mask.sum()), "score < 1 or score > 5", "Conflicts with the documented scale"],
    ["Resolved exceeds received", int(resolved_exceeds_received_mask.sum()), "resolved > received", "Creates an impossible relationship under the stated meaning"],
]
issue_register = pd.DataFrame(
    issue_rows,
    columns=["issue", "affected_rows", "detection_rule", "why_follow_up"]
)
display(issue_register)

print("\nRaw city labels")
print(df["city"].value_counts(dropna=False))
print("\nRaw service labels")
print(df["service_type"].value_counts(dropna=False))

print("\nRows involved in duplicate or identifier review")
display(df.loc[exact_duplicate_mask | repeated_id_mask])
print("\nOwner decisions are required for repeated identifiers, category mappings and the meaning of invalid values.")

Missing values by column
record_id              0
report_month           0
city                   0
service_type           0
cases_received         0
cases_resolved         0
resolution_days        0
satisfaction_score     4
digital_cases          0
in_person_cases        0
repeat_contacts        0
complaints_received    0
staff_hours            0
feedback               3
dtype: int64
                               issue  affected_rows                detection_rule                                                why_follow_up
0               Missing satisfaction              4                        isna()                                Changes the score denominator
1                   Missing feedback              3                        isna()                               Reduces text-analysis coverage
2      Exact duplicate participation              2        duplicated(keep=False)                         May double-count the same stored row
3  Repeated identifier participation    

## Solution 3.14.13 — Examine feedback about waiting and clarity

Run this complete block, inspect the displayed rows and summaries, and compare the final interpretation with your own wording.


In [63]:
import pandas as pd  # Imports pandas so the text exercise runs independently.

data_url = "https://raw.githubusercontent.com/asmrabbi/E26_TAN7_Scripting_CPH/main/data/E26_TAN7_service_experience_raw.csv"  # Stores the raw dataset URL.
df = pd.read_csv(data_url)  # Loads the complete raw dataset again.
assert df.shape == (121, 14), "The wrong dataset version was loaded."

feedback = df["feedback"].dropna()
print("Non-missing feedback comments:", feedback.count())
display(feedback.sample(10, random_state=42))

df["feedback_characters"] = df["feedback"].str.len()
df["feedback_words"] = df["feedback"].str.split().str.len()
display(df[["feedback_characters", "feedback_words"]].describe().round(2))

df["mentions_wait_or_delay"] = df["feedback"].str.contains(
    "wait|long|delay", case=False, na=False, regex=True
)
df["mentions_clarity_or_help"] = df["feedback"].str.contains(
    "clear|helpful|useful", case=False, na=False, regex=True
)

for indicator in ["mentions_wait_or_delay", "mentions_clarity_or_help"]:
    print(f"\n{indicator}: {int(df[indicator].sum())} matching comments")
    display(df.loc[df[indicator], ["record_id", "service_type", "satisfaction_score", "feedback"]])
    print("Service-type comparison")
    display(pd.crosstab(df["service_type"], df[indicator]))
    print("Satisfaction comparison")
    display(df.groupby(indicator)["satisfaction_score"].agg(["count", "mean", "median"]).round(2))
    scored_matches = df.loc[df[indicator], "satisfaction_score"].count()
    print("Matches with a stored satisfaction score:", int(scored_matches))

print("\nInterpretation")
print("These are transparent keyword-rule matches, not complete qualitative themes.")
print("Different wording and negation can create false negatives or false positives.")
print("Any score difference is descriptive association and does not show that the matched language caused the score.")

Non-missing feedback comments: 118
57     The answer was useful, although the waiting ti...
91                    Fast response and useful guidance.
4      The response was fast and the information was ...
46     The response was quick, but route-change infor...
27     The guidance was clear but the process require...
119    The response was fast and the information was ...
70     The answer solved the problem, but repeated co...
10     The online application was useful, but several...
92     The guidance was clear but the process require...
74                    Friendly staff and a clear answer.
       feedback_characters  feedback_words
count               118.00          118.00
mean                 63.78            9.53
std                  14.87            2.10
min                  34.00            5.00
25%                  53.00            8.00
50%                  65.00            9.50
75%                  74.00           11.75
max                  91.00           14.00

mention

## Solution 3.14.14 — Test a strong management claim before repeating it

Run this complete block, inspect the displayed rows and summaries, and compare the final interpretation with your own wording.


In [64]:
import pandas as pd  # Imports pandas for a standalone review of the claim.

data_url = "https://raw.githubusercontent.com/asmrabbi/E26_TAN7_Scripting_CPH/main/data/E26_TAN7_service_experience_raw.csv"  # Records the exact source used in the review.
df = pd.read_csv(data_url)  # Loads a fresh copy of the raw file.
assert df.shape == (121, 14), "The wrong dataset version was loaded."

print("Raw city labels")
print(df["city"].value_counts(dropna=False))
print("\nRaw service labels")
print(df["service_type"].value_counts(dropna=False))

df["mentions_wait"] = df["feedback"].str.contains(
    "wait|long|delay", case=False, na=False, regex=True
)
focal_mask = (df["city"] == "Copenhagen") & (df["service_type"] == "Transport")
focal = df.loc[focal_mask].copy()

print("\nExact Copenhagen–Transport result")
print("Rows:", len(focal))
print("Received-case total:", focal["cases_received"].sum())
print("Satisfaction count:", focal["satisfaction_score"].count())
print("Satisfaction mean:", round(focal["satisfaction_score"].mean(), 2))
print("Satisfaction median:", focal["satisfaction_score"].median())
print("Waiting-rule matches:", int(focal["mentions_wait"].sum()))

exact_cities = ["Copenhagen", "Aarhus", "Odense", "Aalborg"]
exact_services = ["Transport", "Employment", "Waste", "Citizen Services", "Housing"]
comparison = df[
    df["city"].isin(exact_cities) & df["service_type"].isin(exact_services)
].copy()

comparison_summary = comparison.groupby(["city", "service_type"]).agg(
    row_count=("record_id", "count"),
    cases_received_total=("cases_received", "sum"),
    satisfaction_n=("satisfaction_score", "count"),
    satisfaction_mean=("satisfaction_score", "mean"),
    waiting_matches=("mentions_wait", "sum")
)
display(comparison_summary.round(2))

invalid_satisfaction = (
    focal["satisfaction_score"].isna()
    | (focal["satisfaction_score"] < 1)
    | (focal["satisfaction_score"] > 5)
)
impossible_cases = focal["cases_resolved"] > focal["cases_received"]
print("\nFocal quality checks")
print("Missing/out-of-range satisfaction rows:", int(invalid_satisfaction.sum()))
print("Resolved greater than received rows:", int(impossible_cases.sum()))
print("Exact duplicate rows in focal subset:", int(focal.duplicated().sum()))
print("Exact duplicate rows in full data:", int(df.duplicated().sum()))

wait_comments = focal.loc[focal["mentions_wait"], ["record_id", "satisfaction_score", "feedback"]]
display(wait_comments)
print("Focal wait matches:", len(wait_comments))
print("Focal wait matches with a score:", int(wait_comments["satisfaction_score"].count()))

print("\nEvidence decision")
print("Observation: the printed exact-label summaries show the workload, satisfaction and waiting-rule results for each compared group.")
print("Cautious interpretation: the focal group can be compared descriptively after naming its row and score denominators.")
print("Unsupported claim: these synthetic records cannot show that waiting caused low satisfaction in a real population.")
print("Decision: revise the slide so it reports exact observations and removes the causal statement.")

Raw city labels
city
Aarhus          30
Copenhagen      29
Odense          29
Aalborg         28
 copenhagen      1
AALBORG          1
Odense           1
aarhus           1
Koebenhavn       1
Name: count, dtype: int64

Raw service labels
service_type
Transport              24
Employment             24
Waste                  24
Citizen Services       24
Housing                21
Transport               1
Citizen services        1
WASTE                   1
Employment Services     1
Name: count, dtype: int64

Exact Copenhagen–Transport result
Rows: 5
Received-case total: 634
Satisfaction count: 4
Satisfaction mean: 4.0
Satisfaction median: 3.95
Waiting-rule matches: 0
                             row_count  cases_received_total  satisfaction_n  satisfaction_mean  waiting_matches
city       service_type                                                                                         
Aalborg    Citizen Services          6                   612               6               4.60     

# Checks and next steps

All 39 website exercises are present in the student section and in the same solution order. Revisit any answer you could run but could not explain. Examples 3.14.1–3.14.8 remain in the separate Tutorial 3.14 Guided Examples notebook.
